File: Source/IDE/src/Project/Lexer/Tokenizer.cpp
/*
 *  Tokenizer.cpp
 *  Tokenizer
 */

#include <QDebug>
#include <QFile>
#include <Project/Lexer/Tokenizer.hpp>
#include <Project/Lexer/StringStream.hpp>

namespace Lexer
{
	TokenizerSingleton::TokenizerSingleton()
		: delimiters()
	{
		delimiters.push_back(">>=");
		delimiters.push_back("<<=");
		delimiters.push_back("+=");
		delimiters.push_back("-=");
		delimiters.push_back("*=");
		delimiters.push_back("/=");
		delimiters.push_back("%=");
		delimiters.push_back("|=");
		delimiters.push_back("^=");
		delimiters.push_back("&=");
		delimiters.push_back("==");
		delimiters.push_back("!=");
		delimiters.push_back("<=");
		delimiters.push_back(">=");
		delimiters.push_back("<<");
		delimiters.push_back(">>");
		delimiters.push_back("&&");
		delimiters.push_back("||");
		delimiters.push_back("++");
		delimiters.push_back("--");
		delimiters.push_back("..");
		delimiters.push_back("~");
		delimiters.push_back("!");
		delimiters.push_back("%");
		delimiters.push_back("^");
		delimiters.push_back("&");
		delimiters.push_back("*");
		delimiters.push_back("(");
		delimiters.push_back(")");
		delimiters.push_back("-");
		delimiters.push_back("+");
		delimiters.push_back("=");
		delimiters.push_back("[");
		delimiters.push_back("]");
		delimiters.push_back("{");
		delimiters.push_back("}");
		delimiters.push_back("|");
		delimiters.push_back(":");
		delimiters.push_back(";");
		delimiters.push_back("?");
		delimiters.push_back("/");
		delimiters.push_back("<");
		delimiters.push_back(">");
		delimiters.push_back(",");
		delimiters.push_back(".");
	}

	bool TokenizerSingleton::ParseString(const QString& value, QList<Token>& tokens)
	{
		StringStream stream(value);

		while(!stream.IsEos())
		{
			SkipWhitespace(stream);

			if (SkipComment(stream))
				continue;

			if (ParseStringInner(stream, tokens))
				continue;

			if (ParseWord(stream, tokens))
				continue;

			if (ParseNumber(stream, tokens))
				continue;

			if (ParseDelimiter(stream, tokens))
				continue;

			if (stream.IsEos())
				break;

			qDebug() << "Lexer error";
			return false;
		}

		return true;
	}

	bool TokenizerSingleton::ParseFile(const QString& filename, QList<Token>& tokens)
	{
		QFile file(filename);
		if (!file.open(QFile::ReadOnly | QFile::Text))
			return false;

		return ParseString(file.readAll(), tokens);
	}

	void TokenizerSingleton::SkipWhitespace(StringStream& stream)
	{
		while(!stream.IsEos() && QChar(stream.Peek()).isSpace())
			stream.Read();
	}

	bool TokenizerSingleton::SkipComment(StringStream& stream)
	{
		// Skip single line comments
		if (stream.Peek() == '/' && stream.Peek(1) == '/')
		{
			while(!stream.IsEos() && stream.Peek() != '\n')
				stream.Read();

			return true;
		}

		// Skip multiline comments
		if (stream.Peek() == '/' && stream.Peek(1) == '*')
		{
			while(!stream.IsEos() && !(stream.Peek() == '*' && stream.Peek(1) == '/'))
				stream.Read();

			stream.position += 2;
			return true;
		}

		return false;
	}

	bool TokenizerSingleton::ParseStringInner(StringStream& stream, QList<Token>& tokens)
	{
		if (stream.Peek() != '\"')
			return false;

		QString value = "";

		while(stream.Peek() == '\"')
		{
			char ch;

			++stream.position;
			while((ch = stream.Read()) != '\"')
			{
				if (stream.IsEos())
					qDebug() << "Unterminated string";

				value += ch;
			}

			SkipWhitespace(stream);
		}

		tokens.push_back(Token(TokenType::String, value));
		return true;
	}

	bool TokenizerSingleton::ParseWord(StringStream& stream, QList<Token>& tokens)
	{
		if (QChar(stream.Peek()).isLetter() || stream.Peek() == '_')
		{
			QString word = "";

			while(!stream.IsEos() && (QChar(stream.Peek()).isLetterOrNumber() || stream.Peek() == '_'))
				word += stream.Read();

			tokens.push_back(Token(TokenType::Word, word));
			return true;
		}

		return false;
	}

	bool TokenizerSingleton::ParseNumber(StringStream& stream, QList<Token>& tokens)
	{
		if (QChar(stream.Peek()).isDigit())
		{
			QString number = "";

			if (stream.Peek() == '0' && stream.Peek(1) == 'x')
			{
				stream.position += 2;

				while(!stream.IsEos() && IsHex(stream.Peek()))
					number += stream.Read();

				bool ok;
				qlonglong result = number.toLongLong(&ok, 16);
				if (!ok)
					qDebug() << "Invalid hex number";

				number = QString::number(result);
			}
			else if (stream.Peek() == '0' && stream.Peek(1) == 'b')
			{
				stream.position += 2;

				while(!stream.IsEos() && (stream.Peek() == '0' || stream.Peek() == '1'))
					number += stream.Read();

				bool ok;
				qlonglong result = number.toLongLong(&ok, 2);
				if (!ok)
					qDebug() << "Invalid hex number";

				number = QString::number(result);
			}
			else
			{
				while(!stream.IsEos() && (QChar(stream.Peek()).isDigit() || (stream.Peek() == '.' && stream.Peek(1) != '.')))
					number += stream.Read();
			}

			tokens.push_back(Token(TokenType::Number, number));
			return true;
		}

		return false;
	}

	bool TokenizerSingleton::ParseDelimiter(StringStream& stream, QList<Token>& tokens)
	{
		bool foundDelimiter = false;
		QString peekedDelimiter = "";

		for(QString delimiter : delimiters)
		{
			peekedDelimiter = "";

			for(int i = 0; i < delimiter.length(); ++i)
				peekedDelimiter += stream.Peek(i);

			if (peekedDelimiter == delimiter)
			{
				stream.position += delimiter.length();
				foundDelimiter = true;
				break;
			}
		}

		if (foundDelimiter)
		{
			tokens.push_back(Token(TokenType::Delimiter, peekedDelimiter));
			return true;
		}

		return false;
	}

	bool TokenizerSingleton::IsHex(char ch) const
	{
		QChar value = QChar(ch).toUpper();
		return value.isDigit() || (value >= 'A' && value <= 'F');
	}

	TokenizerSingleton& TokenizerSingleton::GetInstance()
	{
		static TokenizerSingleton instance;
		return instance;
	}
}

File: Source/Compiler/src/Source/Lexer/Tokenizer.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;

namespace Swift.Source.Lexer
{
	internal static class Tokenizer
	{
		// NOTE Make sure that the largest delimiters come first in the list
		private static readonly List<string> Delimiters = new List<string>
		{
			"<<=", ">>=",
			"+=", "-=", "*=", "/=", "%=", "|=", "^=", "&=", "==", "!=",
			"<=", ">=", "<<", ">>", "&&", "||", "++", "--", "..",
			"~", "!", "%", "^", "&", "*", "(", ")", "-", "+", "=", "[",
			"]", "{", "}" ,"|", ":", ";", "?", "/", "<", ">", ",", "."
		};

		public static string Filename { get; private set; }
		public static uint Line { get; private set; }

		public static List<Token> ParseFile(string filename)
		{
			Filename = filename;

			try
			{
				using (StreamReader reader = new StreamReader(Filename))
					return ParseStream(new StringStream(reader.ReadToEnd()));
			}
			catch (IOException)
			{
				Console.WriteLine("Cannot open file \"{0}\" for compiling.", Filename);
				throw;
			}
		}

		private static List<Token> ParseStream(StringStream stream)
		{
			List<Token> tokens = new List<Token>();
			Line = 1;

			while (!stream.IsEndOfStream)
			{
				SkipWhitespace(stream);

				if (SkipComment(stream))
					continue;

				if (ParseString(stream, tokens))
					continue;

				if (ParseWord(stream, tokens))
					continue;

				if (ParseNumber(stream, tokens))
					continue;

				if (ParseDelimiter(stream, tokens))
					continue;

				if (stream.IsEndOfStream)
					break;

				throw new LexerException("Unexpected token '{0}' on line {1}.", stream.Read(), Line);
			}

			tokens.Add(new Token(TokenType.EndOfStream, Filename, Line));

			return tokens;
		}

		private static void SkipWhitespace(StringStream stream)
		{
			while (!stream.IsEndOfStream && char.IsWhiteSpace(stream.Peek()))
			{
				// Read new lines
				if (stream.Read() == StringStream.NewLine)
					++Line;
			}
		}

		private static bool SkipComment(StringStream stream)
		{
			// Skip single line comments
			if (stream.Peek() == '/' && stream.Peek(1) == '/')
			{
				while (!stream.IsEndOfStream && stream.Peek() != StringStream.NewLine)
					stream.Read();

				return true;
			}

			// Skip multiline comments
			if (stream.Peek() == '/' && stream.Peek(1) == '*')
			{
				stream.Position += 2;
				int nestCount = 1;

				while (nestCount > 0)
				{
					if (stream.IsEndOfStream)
						throw new LexerException("Unterminated multi-line comment ran past end of file.");

					if (stream.Peek() == '/' && stream.Peek(1) == '*')
						++nestCount;
					else if (stream.Peek() == '*' && stream.Peek(1) == '/')
						--nestCount;
					else
					{
						++stream.Position;
						continue;
					}

					stream.Position += 2;

					if (stream.Read() == StringStream.NewLine)
						++Line;
				}

				return true;
			}

			return false;
		}

		private static bool ParseString(StringStream stream, ICollection<Token> tokens)
		{
			if (!(stream.Peek() == '@' && stream.Peek(1) == '\"') && stream.Peek() != '\"')
				return false;

			uint startLine = Line;
			string value = "";

			bool escaped = true;
			if (stream.Peek() == '@')
			{
				escaped = false;
				stream.Read();
			}

			while (stream.Peek() == '\"')
			{
				char ch;

				++stream.Position;
				while ((ch = stream.Read()) != '\"')
				{
					if (stream.IsEndOfStream)
						throw new LexerException("Unterminated string on line {0}.", startLine);

					if (ch == StringStream.NewLine)
						++Line;

					value += ch;
					if (ch == '\\')
						value += stream.Read();
				}

				SkipWhitespace(stream);
			}

			value = escaped ? ParseEscapeSequence(value) : value;
			tokens.Add(new Token(TokenType.String, value, Filename, startLine));

			return true;
		}

		private static bool ParseWord(StringStream stream, ICollection<Token> tokens)
		{
			if (char.IsLetter(stream.Peek()) || stream.Peek() == '_')
			{
				string word = "";

				while (!stream.IsEndOfStream && (char.IsLetterOrDigit(stream.Peek()) || stream.Peek() == '_'))
					word += stream.Read();

				switch(word)
				{
					case "true":
						tokens.Add(new Token(TokenType.Number, "1", Filename, Line));
						break;

					case "false":
						tokens.Add(new Token(TokenType.Number, "0", Filename, Line));
						break;

					default:
						tokens.Add(new Token(TokenType.Word, word, Filename, Line));
						break;
				}

				return true;
			}

			return false;
		}

		private static bool ParseDelimiter(StringStream stream, ICollection<Token> tokens)
		{
			bool foundDelimiter = false;
			string peekedDelimiter = "";

			foreach (string delimiter in Delimiters)
			{
				peekedDelimiter = "";

				for (int i = 0; i < delimiter.Length; ++i)
					peekedDelimiter += stream.Peek(i);

				if (peekedDelimiter == delimiter)
				{
					stream.Position += delimiter.Length;
					foundDelimiter = true;
					break;
				}
			}

			if (foundDelimiter)
			{
				tokens.Add(new Token(TokenType.Delimiter, peekedDelimiter, Filename, Line));
				return true;
			}

			return false;
		}

		private static bool ParseNumber(StringStream stream, ICollection<Token> tokens)
		{
			if (char.IsDigit(stream.Peek()))
			{
				string number = "";

				if (stream.Peek() == '0' && stream.Peek(1) == 'x')
				{
					stream.Position += 2;

					while (!stream.IsEndOfStream && stream.Peek().IsHex())
						number += stream.Read();

					ulong result;
					try
					{
						result = ulong.Parse(number, NumberStyles.HexNumber, CultureInfo.InvariantCulture);
					}
					catch (OverflowException)
					{
						throw new LexerException("Number is larger than 64bits on line {0}.", Line);
					}
					catch (Exception)
					{
						throw new LexerException("Invalid number on line {0}.", Line);
					}

					number = result.ToString("G");
				}
				else if (stream.Peek() == '0' && stream.Peek(1) == 'b')
				{
					stream.Position += 2;

					while (!stream.IsEndOfStream && char.IsDigit(stream.Peek()))
						number += stream.Read();

					long result;
					try
					{
						result = Convert.ToInt64(number, 2);
					}
					catch(OverflowException)
					{
						throw new LexerException("Number is larger than 64bits on line {0}.", Line);
					}
					catch(Exception)
					{
						throw new LexerException("Invalid number on line {0}.", Line);
					}

					number = result.ToString("G");
				}
				else
				{
					while (!stream.IsEndOfStream && (char.IsDigit(stream.Peek()) || (stream.Peek() == '.' && stream.Peek(1) != '.')))
						number += stream.Read();
				}

				tokens.Add(new Token(TokenType.Number, number, Filename, Line));

				return true;
			}

			return false;
		}

		private static string ParseEscapeSequence(string value)
		{
			StringStream stringStream = new StringStream(value);
			string parsedValue = "";

			while (!stringStream.IsEndOfStream)
			{
				// Read non-escaping characters
				if (stringStream.Peek() != '\\')
				{
					parsedValue += stringStream.Read();
					continue;
				}

				// Skip backslash
				stringStream.Read();

				switch (stringStream.Read())
				{
					case '\'':
						parsedValue += '\'';
						continue;

					case '\"':
						parsedValue += '\"';
						continue;

					case '\\':
						parsedValue += '\\';
						continue;

					case '%':
						parsedValue += '%';
						continue;

					case ':':
						parsedValue += ':';
						continue;

					case '\0':
						parsedValue += '\0';
						continue;

					case 'a':
						parsedValue += '\a';
						continue;

					case 'b':
						parsedValue += '\b';
						continue;

					case 'f':
						parsedValue += '\f';
						continue;

					case 'n':
						parsedValue += '\n';
						continue;

					case 'r':
						parsedValue += '\r';
						continue;

					case 't':
						parsedValue += '\t';
						continue;

					case 'u':
					case 'U':
					case 'x':
					{
						string unicodeValue = "";

						for(int i = 0; i < 8; ++i)
						{
							char read = stringStream.Read();
							unicodeValue += read;

							if (!stringStream.Peek(1).IsHex() && (i == 1 || i == 3 || i == 7))
								break;
						}

						uint unicodeIntValue;
						if (!uint.TryParse(unicodeValue, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out unicodeIntValue))
							throw new LexerException("Malformed escape code \\x, \\u or \\U on line {0}.", Line);

						parsedValue += Convert.ToChar(unicodeIntValue);
						continue;
					}

					case 'v':
						parsedValue += '\v';
						continue;
				}

				throw new LexerException("Bad escape sequence.");
			}

			return parsedValue;
		}

		private static bool IsHex(this char value)
		{
			return char.IsDigit(value) || new HashSet<char>
			{
				'A',
				'B',
				'C',
				'D',
				'E',
				'F'
			}.Contains(char.ToUpper(value, CultureInfo.InvariantCulture));
		}
	}
}
